### PR TITLE
feat(text-input): Add LLM verification after probe undo operation

### PR DESCRIPTION
**Pull request by HuangHan96**
## Overview

This PR adds a secondary LLM verification mechanism in the `enter_text` probe phase to ensure the probe character 'a' is properly removed after the undo operation.

## Problem

Previously, the probe phase only used fixed undo shortcuts (Cmd+Z or Ctrl+Z) to cancel the typed probe character. In some cases, the undo operation may not completely remove the character, leaving it in the input field.

## Solution

After sending the undo keys, add an LLM-based visual verification process to check if the probe character is still visible. If it remains, send a backspace key to delete it.

### Modified Files
- **text_input_vision.go**
  - Added `VerifyProbeCleanup` method to `textInputProbeVision` interface
  - Implemented `llmTextInputVision.VerifyProbeCleanup` with LLM visual analysis
  - Added `buildTextInputProbeCleanupPrompt` to generate verification prompt
  - Added stub implementation for testing

- **text_input_engine.go**
  - Enhanced `probeTextInputMode` defer logic with verification workflow
  - Added logging for verification results
  - Send backspace if probe character is still visible

- **tools_enter_text_test.go**
  - Added `probeVerifyingVision` test helper structure
  - Added test for backspace being sent when character is still visible
  - Added test for backspace being skipped when character is removed

## Workflow

1. Type probe character "a"
2. Wait and capture screenshot for input mode analysis
3. Send undo shortcut (Cmd+Z or Ctrl+Z)
4. **[NEW]** Wait for undo to take effect
5. **[NEW]** Capture screenshot and call LLM to verify if probe character is still visible
6. **[NEW]** If character is still visible, send backspace to delete it
7. **[NEW]** Log verification result

## Testing

All text input related tests pass:
- ✅ New verification tests pass
- ✅ All existing TextInput and EnterText tests pass
- ✅ Successful compilation with no errors

This improvement enhances probe cleanup reliability across different systems and IME environments.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved text-entry cleanup by comparing before-and-after states to distinguish temporary probe characters from pre-existing content.
  * Backspace cleanup now occurs only when the probe is confirmed visible and safe to remove.
  * Cleanup remains optional when baseline capture or vision-based verification is unavailable.
  * Cleanup verification and screenshot failures are now reported instead of silently ignored.

* **Tests**
  * Added coverage confirming undo occurs before backspace cleanup.
  * Expanded coverage for probe visibility and cleanup safety.

### CODE REVIEW

**Comment by aiden-lightning** [comment]
**Automated Claude Code review**

- PR: #637 — feat(text-input): Add LLM verification after probe undo operation
- Head SHA: `cd4c242720ec83bdaf1c934aa7d2fc2df3dfb954`
- GitHub updatedAt: `2026-09-04T03:29:22Z`
- Reviewed at: `2026-09-04T11:51:38+08:00`

## PR Review: feat(text-input): Add LLM verification after probe undo operation

### Summary

This PR adds LLM-based verification after the probe undo operation in text input mode detection. When undo fails to remove the probe character "a", the system now detects this and sends a backspace as fallback cleanup. The implementation is well-structured with proper error handling, optional interface support, and comprehensive test coverage.

### Verified Issues

**None found.** After thorough examination of the code changes, I verified the following aspects and found them to be correct:

### Observations

**1. Architecture - Optional Interface (✓ Correct)**
- The `textInputProbeCleanupVision` interface is optional by design
- `cleanupSupported` check at `text_input_engine.go:317` gracefully handles visions that don't implement cleanup
- This allows existing stub implementations to continue working without modification

**2. Screenshot Timing (✓ Correct)**
- "Before" screenshot: captured at `text_input_engine.go:321` before typing "a"
- "After" screenshot: captured at `text_input_engine.go:354` after undo+wait
- Timing matches the LLM prompt expectations in `buildTextInputProbeCleanupPrompt`

**3. Defer Logic with Named Returns (✓ Correct)**
- `vlmCalls++` at line 362 correctly increments the named return value inside defer
- Go's defer semantics ensure this increment affects the returned value
- Error handling uses `errors.Join()` appropriately to preserve all errors

**4. Safety Logic (✓ Correct)**
- At `text_input_vision.go:220-222`: Returns error when `probe_character_visible=true` AND `cleanup_safe=false`
- This prevents potentially destructive backspace when we can't confirm the visible "a" is from the probe
- Conservative approach is appropriate for avoiding data loss

**5. Test Coverage (✓ Comprehensive)**
- Unit tests verify JSON parsing edge cases (missing fields, unsafe cleanup)
- Integration tests verify both code paths (backspace sent vs skipped)
- Mock implementation `probeVerifyingVision` properly simulates the interface

**6. Multi-Screenshot Vision API (✓ Well Designed)**
- `visionJSONWithScreenshots` at `text_input_vision.go:497` properly constructs multi-image prompt
- Validates each screenshot has data before adding to message parts
- Single-screenshot calls work via `visionJSON` wrapper for backward compatibility

**7. Error Propagation (✓ Sound)**
- Cleanup errors are logged and joined with main error
- Early returns in defer prevent cascading failures
- Context cancellation handled naturally through error propagation

### Verdict

**LGTM** - The PR is well-implemented with no defects found. The changes are:
- **Isolated**: Only affects probe cleanup path, existing flows unchanged
- **Defensive**: Optional interface ensures backward compatibility
- **Safe**: Conservative error handling prevents data loss
- **Tested**: Both unit and integration tests cover the new logic paths

The implementation correctly addresses the problem of undo not working consistently across platforms by adding visual verification and fallback cleanup.

File: src/agent/internal/agent/text_input_vision.go [hunk 1]
```diff
@@ -67,6 +67,7 @@ type textInputProbeAnalysis struct {
 
 type textInputProbeVision interface {
 	ProbeInputMode(ctx context.Context, screenshot screenshotResult, platform string, focus focusPointArgs) (textInputProbeAnalysis, error)
+	VerifyProbeCleanup(ctx context.Context, screenshot screenshotResult, platform string, focus focusPointArgs) (bool, error)
```
[the remark this answers is not in the document]
Comment (HuangHan96) [reply src/agent/internal/agent/text_input_vision.go]: Addressed in commit f63a471c. ProbeInputMode remains unchanged in textInputProbeVision, and cleanup verification now uses the separate optional textInputProbeCleanupVision interface. probeTextInputMode skips cleanup verification for probe-only vision implementations, so probePlanningConcurrentVision can proceed normally.

File: src/agent/internal/agent/text_input_vision.go [hunk 2]
```diff
@@ -198,6 +199,38 @@ Classification rules:
...
+		return false, err
+	}
+	var parsed struct {
+		ProbeCharacterVisible bool   `json:"probe_character_visible"`
+		Evidence              string `json:"evidence"`
+	}
+	if err := json.Unmarshal([]byte(raw), &parsed); err != nil {
+		return false, fmt.Errorf("parse probe cleanup verification: %w", err)
+	}
+	return parsed.ProbeCharacterVisible, nil
```
[the remark this answers is not in the document]
Comment (HuangHan96) [reply src/agent/internal/agent/text_input_vision.go]: Addressed in commit 4497c08e. probe_character_visible is now presence-aware via *bool; omitted decisions return an explicit error instead of being treated as false. Added TestTextInputProbeCleanupRejectsMissingVisibilityDecision.

File: src/agent/internal/agent/text_input_vision.go [hunk 3]
```diff
@@ -198,6 +199,38 @@ Classification rules:
...
+Focused field location (normalized 0-1000): (%.0f, %.0f)
+
+Return JSON only using exactly this shape:
+{"probe_character_visible":false,"evidence":"short visual reason"}
+
+Verification rules:
+- Inspect the focused input field at the specified coordinates.
+- Set probe_character_visible=true if the lowercase letter "a" is still visible in the field as either committed text or uncommitted preedit text.
+- Set probe_character_visible=false if the field is empty or the "a" has been successfully removed.
+- Ignore any "a" characters that were already in the field before the probe operation.
```
[the remark this answers is not in the document]
Comment (HuangHan96) [reply src/agent/internal/agent/text_input_vision.go]: Addressed in commit 789062ba. Cleanup verification now receives both the pre-probe baseline screenshot and the post-undo screenshot, and the prompt requires confirming cleanup safety before any destructive fallback.

File: src/agent/internal/agent/tools_enter_text_test.go [hunk 4]
```diff
@@ -295,6 +295,111 @@ func TestTextInputProbeSupportsWindowsAndLinuxPlatforms(t *testing.T) {
...
+
+	// Verify that backspace was sent after undo
+	undoIndex, backspaceIndex := -1, -1
+	for index, call := range keyboardTap.calls {
+		if strings.Contains(call, "meta") && strings.Contains(call, "z") {
+			undoIndex = index
+		}
+		if strings.Contains(call, "backspace") {
+			backspaceIndex = index
+		}
```
[the remark this answers is not in the document]
Comment (HuangHan96) [reply src/agent/internal/agent/tools_enter_text_test.go]: Addressed in commit 4497c08e. The test records the last undo index and the first Backspace index, then asserts that undo occurs before the first Backspace.

File: src/agent/internal/agent/text_input_engine.go [hunk 5]
```diff
@@ -314,15 +315,69 @@ func (e *textInputEngine) probeTextInputMode(ctx context.Context, platform strin
...
+			log.Printf("[text-input] probe cleanup verification screenshot failed: %v", captureErr)
+			return
+		}
+
+		// Ask LLM to verify if probe character is still visible
+		probeStillVisible, verifyErr := cleanupVision.VerifyProbeCleanup(ctx, probeBeforeScreenshot, verifyShot, platform, focus)
+		vlmCalls++
+		if verifyErr != nil {
+			log.Printf("[text-input] probe cleanup verification failed: %v", verifyErr)
+			err = errors.Join(err, fmt.Errorf("probe cleanup verification: %w", verifyErr))
```
[the remark this answers is not in the document]
Comment (HuangHan96) [reply src/agent/internal/agent/text_input_engine.go:366]: Addressed in commit 5509e445. Post-undo screenshot failures are now joined into the named probe error, while the fail-safe behavior still avoids sending Backspace.

File: src/agent/internal/agent/text_input_vision.go [hunk 6]
```diff
@@ -198,6 +202,44 @@ Classification rules:
...
+		CleanupSafe           bool   `json:"cleanup_safe"`
+		Evidence              string `json:"evidence"`
+	}
+	if err := json.Unmarshal([]byte(raw), &parsed); err != nil {
+		return false, fmt.Errorf("parse probe cleanup verification: %w", err)
+	}
+	if parsed.ProbeCharacterVisible && !parsed.CleanupSafe {
+		return false, fmt.Errorf("probe character remains visible but cleanup is unsafe")
+	}
+	return parsed.ProbeCharacterVisible, nil
```
[the remark this answers is not in the document]
Comment (HuangHan96) [reply src/agent/internal/agent/text_input_vision.go]: Addressed in commit 4497c08e. A response with probe_character_visible=true and cleanup_safe=false now returns an explicit unsafe-cleanup error, which probeTextInputMode propagates and uses to abort without Backspace.
